### PR TITLE
Fix compiling and linking the php plugin in uwsgi

### LINKED ISSUE #97352: uwsgi with php plugin doesn't compile
**Describe the bug**
when enabling the php plugin in the uwsgi package the compile of the plugin fails.

**To Reproduce**
1. `nix-build -E 'with import <nixpkgs> {}; uwsgi.override { plugins=[ "php" ]; }'`

**Screenshots**
```
*** uWSGI is ready, launch it with ./uwsgi ***
using profile: buildconf/nixos.ini
detected include path: ['/nix/store/vb1fh8zqdcs7n3hvn8awlrpdfnic6zjg-gcc-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0/include', '/nix/store/vb1fh8zqdcs7n3hvn8awlrpdfnic6zjg-gcc-9.3.0/include', '/nix/store/vb1fh8zqdcs7n3hvn8awlrpdfnic6zjg-gcc-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0/include-fixed', '/nix/store/vbs3xx1b5i5bzf1sjkz2lkmz7b8cy92y-glibc-2.30-dev/include']
*** uWSGI building and linking plugin plugins/php ***
/bin/sh: php-config: not found
/bin/sh: php-config: not found
/bin/sh: php-config: not found
/bin/sh: php-config: not found
[gcc] /nix/store/yjg3pdnwgynl1sz9rkc5gyjcgjamgfi9-uwsgi-2.0.18/lib/uwsgi/php_plugin.so
In file included from plugins/php/php_plugin.c:1:
plugins/php/common.h:1:10: fatal error: php.h: No such file or directory
    1 | #include "php.h"
      |          ^~~~~~~
compilation terminated.
In file included from plugins/php/session.c:1:
plugins/php/common.h:1:10: fatal error: php.h: No such file or directory
    1 | #include "php.h"
      |          ^~~~~~~
compilation terminated.
*** unable to build php plugin ***
```

**Additional context**

The uwsgi build system tries to invoke `php-config` to get the include paths for the compile, but php-config is not in the path (it's in the php.dev package). This can be fixed by setting the `UWSGICONFIG_PHPDIR = "${php-embed.unwrapped.dev}";` environment variable. uwsgi then finds `php.h`, but fails on `#include "ext/session/php_session.h"` which is in `extensions.session.dev`, and under a different path in the store.

https://github.com/unbit/uwsgi/blob/master/plugins/php/common.h
https://github.com/unbit/uwsgi/blob/master/plugins/php/uwsgiplugin.py

I'll try to find a better solution next.

**Notify maintainers**
@abbradar @schneefux @globin

**Metadata**
Maintainer information:
```yaml
 - system: `"x86_64-linux"`
 - host os: `Linux 5.8.7-arch1-1, Alpine Linux, noversion`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixpkgs-20.09pre228453.dcb64ea42e6"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by gdamjan**
###### Motivation for this change

fixes #97352 - it enables the use of the php plugin of uwsgi.

###### Things done

I tested in a nixos docker container, on ArchLinux host:

```sh
$ docker run -ti --rm -v ~/src/nixpkgs:/nixpkgs nixos/nix:latest
# nix-build -E 'with import <nixpkgs> {}; uwsgi.override { plugins=[ "php" ]; }' # fails as specified in the bug 97352

# NIX_PATH=nixpkgs=/nixpkgs nix-build -E 'with import <nixpkgs> {}; uwsgi.override { plugins=[ "php" ]; }' # works
```

###### Explanation of the changes

The root of the problem is that uwsgi has a bit of a different build system, and it would expect to find the `php-config` binary in order to set it's compiler and linker flags. There was an additional problem that the uwsgi-php plugin needs to also link with the `session.so` php extension - which on NixOS is a separate package¹.

The first problem was solved by adding all the required `buildInputs` to the uwsgi package so that it can find all the needed header files, libraries and to have php-config in its PATH.

I also had to patch uwsgi to **not** look for `php_session.h` with a path like this `#include "ext/session/php_session.h"`, since on NixOS it's just in `…php-session-7.4.8-dev/include`.

Third, I had to patch the uwsgi build system for the php plugin to explicitly add the `session.so` library and its rpath to `LDFLAGS`. In the nix expression that's exported as the UWSGICONFIG_PHP_LDFLAGS environment variable. It's important that these flags are not used for the main uwsgi binary, but just for the php-plugin shared object.

¹ not sure why the session extension is separated on NixOS, it's builtin in libphp.so on both Debian/Ubuntu and ArchLinux - something to be discussed by the PHP team i guess.

ps.
this is my first involvement with the NixOS project, so I'm open to any criticism and suggestions.

### CODE REVIEW

File: pkgs/servers/uwsgi/default.nix
Comment (schneefux) [inline pkgs/servers/uwsgi/default.nix:94]: Does uwsgi still build when the php plugin is not enabled? Doesn't this require php-embed to be built on all systems that install uwsgi?
The patches should be applied conditionally when the PHP plugin is enabled so the uwsgi build doesn't break when php-embed breaks.

**Review by schneefux — COMMENTED** [review]
Thank you for your contribution! Also, great job on the research 👍
Maybe someone with more uwsgi/Nix experience like @Mic92 can also have a look.

File: pkgs/servers/uwsgi/default.nix
Comment (gdamjan) [reply pkgs/servers/uwsgi/default.nix:94]: good point. let me double check this.
thanks for reviewing.

Comment (gdamjan) [reply pkgs/servers/uwsgi/default.nix:94]: - uwsgi builds even without the php plugin enabled
- the patches are just to the parts of the uwsgi build system for the php plugin itself, so they don't change the uwsgi build process at all
- you are correct that setting UWSGICONFIG_PHP_LDFLAGS unconditionally pulls php-embed as a dependency to uwsgi even if php is not used. I'll fix this.

Comment (gdamjan) [reply pkgs/servers/uwsgi/default.nix]: fixed in b55cc92f5cc81d47e68330b85d71ce7422b175a6

**Review by schneefux — APPROVED** [review]

**Comment by gdamjan** [comment]
Should I prepare a PR for other branches of nixpkgs too?

**Comment by schneefux** [comment]
Let's wait until this is merged (I can only review)

**Comment by gdamjan** [comment]
can we get some reviews here please? :)

**Comment by gdamjan** [comment]
btw, I have this change in my tree too. not sure if I should add it to this PR, or to open a new one:

```
Date:   Thu Sep 10 21:03:19 2020 +0200

    use systemd.lib if it exists or just systemd

diff --git a/pkgs/servers/uwsgi/default.nix b/pkgs/servers/uwsgi/default.nix
index 043b788673b..9ede8249f1c 100644
--- a/pkgs/servers/uwsgi/default.nix
+++ b/pkgs/servers/uwsgi/default.nix
@@ -71,7 +71,8 @@ stdenv.mkDerivation rec {
 
   buildInputs =  [ jansson pcre ]
               ++ lib.optional withPAM pam
-              ++ lib.optional withSystemd systemd
+              ++ lib.optional withSystemd (lib.getLib systemd)
+              ++ lib.optional withSystemd (lib.getDev systemd)
               ++ lib.concatMap (x: x.inputs) needed
               ;
 
```

**Comment by schneefux** [comment]
Hmmm, I tested this on my server and all my web pages are blank. Could these changes have caused it?

**Comment by gdamjan** [comment]
> I tested this on my server and all my web pages are blank. Could these changes have caused it?

well… more context is needed.

Prior to this PR you couldn't even have uwsgi with PHP in nix. It just doesn't compile (see #97352).

Apart from that, I don't know what your web pages are, what your uwsgi configuration is, or how you use it. I can assure you that I'm currently running uwsgi+php and a php app from this exact branch.

**Comment by aanderse** [comment]
@rnhmjoj is this something you have the ability to review?

**Comment by rnhmjoj** [comment]
> @rnhmjoj is this something you have the ability to review?

I think so, I spent some time studying uwsgi recently: I'll see what I can do.

> Apart from that, I don't know what your web pages are, what your uwsgi configuration is, or how you use it. I can assure you that I'm currently running uwsgi+php and a php app from this exact branch.

It would be good to have a small test. I'll try to write it but I know nothing about php.

**Comment by gdamjan** [comment]
> It would be good to have a small test. I'll try to write it but I know nothing about php.

thanks for that

**Comment by rnhmjoj** [comment]
It took some trial and error but I managed to write a minimal test: It works all right. @schneefux may have had a more complicated setup or just misconfigured uwsgi: even if the doc is good I still find it quite tricky to get it right.

Anyway, @gdamjan I had to rebase because your checkout is too old and the test has been changed in master.

File: pkgs/servers/uwsgi/default.nix
Comment (rnhmjoj) [inline pkgs/servers/uwsgi/default.nix:94]: Can you use `stdenv.lib.optionalString` here? Maybe also `concatStringsSep ","` to break up the line.

File: pkgs/servers/uwsgi/default.nix
Comment (rnhmjoj) [inline pkgs/servers/uwsgi/default.nix:68]: Can you break this line? It's to make the diff easier to read in the future.
Also, I would remove the number prefixes.

**Review by rnhmjoj — CHANGES_REQUESTED** [review]
It looks good to me, besides those two nits.
I'm not too fond of patching software but the fault here is on uwsgi, with their bizarre build system.

**Comment by rnhmjoj** [comment]
Thank you, can you squash the fixup commits? Looks good to be merged.

**Comment by gdamjan** [comment]
> Thank you, can you squash the fixup commits? Looks good to be merged.

just the two last commits into a single, or those two into the second one perhaps?
thanks

**Comment by rnhmjoj** [comment]
> just the two last commits into a single, or those two into the second one perhaps?

I was thinking about squashing them into 27e5cbb.

**Comment by gdamjan** [comment]
> > just the two last commits into a single, or those two into the second one perhaps?
> 
> I was thinking about squashing them into [27e5cbb](https://github.com/NixOS/nixpkgs/commit/27e5cbbf083e25cbf2b8199305c23bcf66c89d52).

right. done!

**Comment by rnhmjoj** [comment]
Uh, while you're at it: can you add the test to `passthru.tests`? (See 8ff681a7caeec1b6138f03242823ec95983a066c for an example)
It makes the CI (ofBorg) automatically run the test when the uwsgi package is updated.

**Review by rnhmjoj — APPROVED** [review]
Thank you!

**Comment by aanderse** [comment]
Thank you very much @rnhmjoj! :tada:

**Comment by fabianhjr** [comment]
Hi, just as a heads up I am having an issue with nixpkgs-review after this commit:

```
$ nix-env -f /home/fabian/.cache/nixpkgs-review/pr-107345/nixpkgs -qaP --xml --out-path --show-trace
error: --- TypeError ----------------------------------------------------------------------------------------------------------- nix-env
at: (97:29) in file: /home/fabian/.cache/nixpkgs-review/pr-107345/nixpkgs/pkgs/servers/uwsgi/default.nix

    96|   # the hack works in coordination with ./additional-php-ldflags.patch
    97|   UWSGICONFIG_PHP_LDFLAGS = lib.optionalString (builtins.any (x: x.name == "php") needed)
      |                             ^
    98|         lib.concatStringsSep "," [

attempt to call something which is not a function but a string
```

**Comment by fabianhjr** [comment]
Yeah, seems like eval is borked after this merge, all recent PRs are having an issue during evaluation: #107340, #107342, #107343, #107344, #107345

```
nix-env failed:
warning: ignoring the user-specified setting 'restrict-eval', because it is a restricted setting and you are not a trusted user
error: while evaluating anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/.gc-of-borg-outpaths.nix:44:12, called from undefined position:
while evaluating anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/top-level/release-lib.nix:121:6, called from /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/attrsets.nix:292:43:
while evaluating 'hydraJob' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/customisation.nix:171:14, called from /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/top-level/release-lib.nix:121:14:
while evaluating the attribute 'drvPath' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/customisation.nix:188:13:
while evaluating the attribute 'drvPath' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/customisation.nix:155:13:
while evaluating the attribute 'buildInputs' of the derivation 'cutelyst-2.13.0' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/stdenv/generic/make-derivation.nix:192:11:
while evaluating the attribute 'UWSGICONFIG_PHP_LDFLAGS' of the derivation 'uwsgi-2.0.19.1' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/stdenv/generic/make-derivation.nix:192:11:
attempt to call something which is not a function but a string, at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/servers/uwsgi/default.nix:97:29
```

**Comment by vcunat** [comment]
Reverted.  Please we do have CI that catches such errors.  There's this red cross next to the commit.  https://gist.github.com/GrahamcOfBorg/cff3c8276a4df394f26fb8693825f8b6

**Comment by rnhmjoj** [comment]
Oof, sorry. I've tested locally by running the test but it must have been cached...

**Comment by gdamjan** [comment]
can someone explain that error, and how to see it locally?

**Comment by rnhmjoj** [comment]
I think you're missing parenthesis around the call to `concatStringsSep`.

**Comment by vcunat** [comment]
Caching won't prevent evaluation errors.

@gdamjan: anything with that package will fail, e.g.
```
$ nix-build -A uwsgi
error: attempt to call something which is not a function but a string, at /home/admin/nix/nixpkgs-x/pkgs/servers/uwsgi/default.nix:97:29
```

**Comment by gdamjan** [comment]
What's the process now. Open a new PR?

**Comment by rnhmjoj** [comment]
> What's the process now. Open a new PR?

Yes, I don't think there is any other way with github.

**Comment by gdamjan** [comment]
> @gdamjan: anything with that package will fail, e.g.

right, for some reason php-opcache fails to compile on the latests master (for me), and I didn't catch the my error.

**Pull request merged**
